# Homework 3  - Tryout pandas and numpy

Copy this notebook. Rename it as: YOURNAME-Homework-panda-numpy-Sept-2017

with your name replacing yourname and  replaced with the date you submit or to the last part.

Do the homeworks problems below.

Upload your completed jupyter notebook to your github site and send me the url via the elearning site as your homework submission. Please note that you need to do your own 3.4, and 3.5 if you share data and analysis from others then you cannot get more than a 3.  

## Homework 3.1

### 3.1.a
Load the data from: http://opendata.dc.gov/datasets that I have include in this github   
into a dataframe.  ( The file has been is available in directory ./data/ccp_current_csv.csv )

In [31]:
import pandas as pd
data_frame=pd.read_csv("C:\Users\AsusGo\Desktop\ccp_current_csv.csv")
data_frame.head(5)


STREET QUADRANT  FROMINTERSECTION              TOINTERSECTION  \
0           13th St       NW        Kennedy St               Longfellow St   
1              S St       NW  North Capitol St                      3rd St   
2  Rhode Island Ave       NE  Brentwood Rd (W)            Brentwood Rd (M)   
3           Mill Rd       NW           27th St                  Cul De Sac   
4  Rhode Island Ave       NE  Brentwood Rd (M)  14th St  Evarts St  Montan   

   NOOFBLOCKS                 WORKDESCRIPTION  YEARBUDGETED  PERCENTCOMPLETED  \
0           1  Pavement Resurface/restoration          2004               100   
1           3                   Regular Cover          1999               100   
2           1  Pavement Resurface/restoration          2004               100   
3           1  Pavement Resurface/restoration          2004               100   
4           2  Pavement Resurface/restoration          2005               100   

      STATUS  MILES      ...             YCOORD   LATITUDE  LONGITUDE  \
0  Completed   0.07      ...       143205.11981  38.956747 -77.029694   
1  Completed   0.33      ...       138468.90256  38.914085 -77.012411   
2  Completed   0.09      ...       139574.10659  38.924041 -76.987221   
3  Completed   0.07      ...       138132.90198  38.911046 -77.054077   
4  Completed   0.16      ...       139657.25318  38.924790 -76.985068   

           PROJECTNOTE                                        PROJECTNAME  \
0                  NaN                                             FedAid   
1                  NaN                                                NaN   
2                  NaN           Fedral Aid Citywide Pavement Restoration   
3  POKA-2003-B-0048-JJ  Citywide Pavement Restoration Contract (Local ...   
4                  NaN                                                NaN   

          PROJECTMGR  PROJECTID        TEAMREP                 EMAIL  \
0    Paul Stephenson       5834        O. Hill                   NaN   
1                NaN       2467            NaN                   NaN   
2  Mohamed Abdullahi       5486   Said Cherifi   said.cherifi@dc.gov   
3    Maduabuchi Udeh       5796  Michael Jelen  Michael.Jelen@dc.gov   
4                NaN       8237            NaN                   NaN   

   CONTACTPHONE  
0  202-671-4591  
1           NaN  
2  202-671-4611  
3  202-671-4542  
4           NaN  

[5 rows x 27 columns]

### 3.1.a  what is its shape and what does that mean?

In [32]:
data_frame.shape

(465, 27)

Shape will let us know number of columns and rows.As it has returned that there are 565 rows and 27 columns in the csv file above.

## Homework 3.2
 What are the number of rows in each 'QUADRANT' ?

In [110]:
q_unique=data_frame.QUADRANT.unique()
print q_unique
NW_count=data_frame.loc[data_frame['QUADRANT']=="NW"].shape
print NW_count
NE_count=data_frame.loc[data_frame['QUADRANT']=="NE"].shape
print NE_count
SW_count=data_frame.loc[data_frame['QUADRANT']=="SW"].shape
print SW_count
SE_count=data_frame.loc[data_frame['QUADRANT']=="SE"].shape
print SE_count
BN_count=data_frame.loc[data_frame['QUADRANT']=="BN"].shape
print BN_count

['NW' 'NE' 'SE' 'SW' 'BN']
(195, 27)
(163, 27)
(11, 27)
(84, 27)
(12, 27)


## Homework 3.3 - Array math demonstration 
For two arrarys  

a= [1,2,3,4]  type=float

b= [5,6,7,8]  type=float

Peform the following array operations using numpy   
( show both operational use of numpy and functional (example addition operation => + vs addition function => numbpy.add() )

### 3.3.1 addition a+b 

In [34]:
a=np.array([1,2,3,4],dtype=np.float)
b=np.array([5,6,7,8],dtype=np.float)
s=np.add(a, b)
print s
a+b

[  6.   8.  10.  12.]


array([  6.,   8.,  10.,  12.])

### 3.3.2 subtraction a-b

In [35]:
su=np.subtract(a,b)
print su
a-b

[-4. -4. -4. -4.]


array([-4., -4., -4., -4.])

### 3.3.3 multiplication a*b

In [36]:
m=np.multiply(a,b)
print m
a*b

[  5.  12.  21.  32.]


array([  5.,  12.,  21.,  32.])

### 3.3.4 divsion a/b

In [37]:
d=np.divide(a,b)
print d
a/b

[ 0.2         0.33333333  0.42857143  0.5       ]


array([ 0.2       ,  0.33333333,  0.42857143,  0.5       ])

### 3.3.5 modulo  a%b

In [38]:
mo=np.mod(a,b)
print mo
a%b

[ 1.  2.  3.  4.]


array([ 1.,  2.,  3.,  4.])

### 3.3.6 power  a^b

In [8]:
pow=np.float_power(a,b)
print pow
a**b

NameError: name 'np' is not defined

## Homework 3.4
Find your own data and load it into a dataframe

In [9]:
import pandas as pd
df=pd.read_csv('C:\Users\AsusGo\Desktop\crime_2015.csv')
df.head(5)

#import numpy as np
#csv = np.genfromtxt('C:\Users\AsusGo\Desktop\crime_2015.csv', delimiter= ",")


X          Y       CCN                REPORT_DAT     SHIFT  METHOD  \
0 -76.979019  38.866897   6155556  2015-12-04T00:00:00.000Z  MIDNIGHT     GUN   
1 -77.022947  38.910389   9051410  2015-09-14T00:00:00.000Z  MIDNIGHT  OTHERS   
2 -76.980393  38.899616  11105327  2015-08-14T00:00:00.000Z  MIDNIGHT   KNIFE   
3 -76.933413  38.890401  11535619  2015-08-30T02:57:29.000Z  MIDNIGHT     GUN   
4 -77.072880  38.930449  14158527  2015-01-25T00:00:00.000Z  MIDNIGHT  OTHERS   

     OFFENSE                                     BLOCK    XBLOCK    YBLOCK  \
0   HOMICIDE       1900 - 1925 BLOCK OF 18TH STREET SE  401821.0  133230.0   
1   HOMICIDE        1500 - 1599 BLOCK OF 8TH STREET NW  398010.0  138058.0   
2  SEX ABUSE      1516 - 1699 BLOCK OF BENNING ROAD NE  401701.0  136862.0   
3    ROBBERY    4800 - 4899 BLOCK OF CENTRAL AVENUE NE  405777.0  135841.0   
4  SEX ABUSE  3000 - 3199 BLOCK OF WISCONSIN AVENUE NW  393681.0  140287.0   

     ...     CENSUS_TRACT VOTING_PRECINCT     XCOORD     YCOORD   LATITUDE  \
0    ...           7605.0    Precinct 112  401795.67  133225.61  38.866890   
1    ...           4901.0     Precinct 18  397984.28  137924.44  38.910382   
2    ...           7901.0     Precinct 81  401716.19  136822.69  38.899608   
3    ...           7804.0     Precinct 98  405802.40  135825.39  38.890393   
4    ...            400.0     Precinct 12  393647.71  140311.55  38.930441   

   LONGITUDE  BID                START_DATE                  END_DATE  \
0 -76.979017  NaN  2006-11-13T19:35:00.000Z  2006-11-13T19:38:00.000Z   
1 -77.022944  NaN  2009-04-18T01:05:00.000Z  2009-04-18T05:00:00.000Z   
2 -76.980391  NaN  2011-07-21T20:00:00.000Z  2011-07-21T23:00:00.000Z   
3 -76.933411  NaN  2015-08-30T02:14:49.000Z  2015-08-30T02:20:15.000Z   
4 -77.072878  NaN  2014-03-21T22:00:00.000Z  2014-03-21T22:05:00.000Z   

   OBJECTID  
0  71503077  
1  71529903  
2  71577169  
3  71588612  
4  71652614  

[5 rows x 26 columns]

## Homework 3.5
Provide an interesting analysis of the data columns ( frequency or averages )

Crimes done at different times

In [10]:
s_unique=df.SHIFT.unique()
print s_unique

['MIDNIGHT' 'DAY' 'EVENING']


In [11]:
MIDNIGHT_count=df.loc[df['SHIFT']=="MIDNIGHT"].shape
print MIDNIGHT_count
EVENING_count=df.loc[df['SHIFT']=="EVENING"].shape
print EVENING_count
DAY_count=df.loc[df['SHIFT']=="DAY"].shape
print DAY_count

(7023, 26)
(16253, 26)
(14042, 26)


According to the above results,it is determined that most of the crime took place in the evening shifts(16253).

In [12]:
o_unique=df.OFFENSE.unique()
print o_unique

['HOMICIDE' 'SEX ABUSE' 'ROBBERY' 'THEFT F/AUTO' 'BURGLARY' 'THEFT/OTHER'
 'MOTOR VEHICLE THEFT' 'ASSAULT W/DANGEROUS WEAPON' 'ARSON']


In [59]:
df.describe()


X             Y           CCN         XBLOCK         YBLOCK  \
count  37318.000000  37318.000000  3.731800e+04   37318.000000   37318.000000   
mean     -77.008079     38.907133  1.512027e+07  399300.066247  137696.929799   
std        0.035877      0.030863  1.077086e+05    3111.622564    3426.116858   
min      -77.113645     38.813478  6.155556e+06  390147.000000  127300.000000   
25%      -77.031963     38.892015  1.505965e+07  397228.000000  136018.067500   
50%      -77.012949     38.906466  1.511258e+07  398876.980000  137623.000000   
75%      -76.985515     38.926430  1.516609e+07  401257.000000  139839.000000   
max      -76.910023     38.993568  1.617462e+07  407806.000000  147292.000000   

               WARD      DISTRICT           PSA  CENSUS_TRACT         XCOORD  \
count  37318.000000  37267.000000  37265.000000  37203.000000   37181.000000   
mean       4.421486      3.695280    374.080907   6210.887321  399303.374851   
std        2.339705      1.947808    194.542204   3145.199626    3110.585803   
min        1.000000      1.000000    101.000000    100.000000  390132.600000   
25%        2.000000      2.000000    206.000000   3400.000000  397217.800000   
50%        5.000000      4.000000    401.000000   7000.000000  398871.560000   
75%        6.000000      5.000000    506.000000   8904.000000  401256.720000   
max        8.000000      7.000000    708.000000  11100.000000  407828.700000   

              YCOORD      LATITUDE     LONGITUDE      OBJECTID  
count   37181.000000  37318.000000  37318.000000  3.731800e+04  
mean   137695.134174     38.907125    -77.008077  7.170944e+07  
std      3425.240318      0.030863      0.035877  5.206550e+04  
min    127319.990000     38.813471    -77.113642  7.150308e+07  
25%    136037.320000     38.892007    -77.031960  7.166899e+07  
50%    137642.700000     38.906458    -77.012947  7.167833e+07  
75%    139856.110000     38.926422    -76.985512  7.177698e+07  
max    147424.260000     38.993560    -76.910021  7.181773e+07

In [60]:
df.groupby("METHOD")["OFFENSE"]

In [37]:
print df.groupby('OFFENSE')['CCN','METHOD','SHIFT'].max()
print df.groupby('OFFENSE')['OBJECTID','METHOD'].min()

                                 CCN  METHOD     SHIFT
OFFENSE                                               
ARSON                       15208488  OTHERS  MIDNIGHT
ASSAULT W/DANGEROUS WEAPON  15703603  OTHERS  MIDNIGHT
BURGLARY                    15831561  OTHERS  MIDNIGHT
HOMICIDE                    15207060  OTHERS  MIDNIGHT
MOTOR VEHICLE THEFT         15208268  OTHERS  MIDNIGHT
ROBBERY                     15888801  OTHERS  MIDNIGHT
SEX ABUSE                   15207028  OTHERS  MIDNIGHT
THEFT F/AUTO                15914474  OTHERS  MIDNIGHT
THEFT/OTHER                 16174617  OTHERS  MIDNIGHT
                            OBJECTID  METHOD
OFFENSE                                     
ARSON                       71663758  OTHERS
ASSAULT W/DANGEROUS WEAPON  71657702     GUN
BURGLARY                    71657701     GUN
HOMICIDE                    71503077     GUN
MOTOR VEHICLE THEFT         71657320     GUN
ROBBERY                     71588612     GUN
SEX ABUSE                   7157716

In [27]:
print df['OFFENSE'].max()
print df['OFFENSE'].min()

THEFT/OTHER
ARSON


In [25]:
print df['SHIFT'].max()
print df['SHIFT'].min()

MIDNIGHT
DAY


In [47]:
print df['OBJECTID'].count()

37318


In [50]:
df['VOTING_PRECINCT'].count()

37305

In [53]:
df.groupby('METHOD')['LATITUDE','LONGITUDE'].max()

LATITUDE  LONGITUDE
METHOD                      
GUN     38.984056 -76.911744
KNIFE   38.980411 -76.910146
OTHERS  38.993560 -76.910021

In [54]:
df.groupby('OFFENSE')['LATITUDE','LONGITUDE'].max()

LATITUDE  LONGITUDE
OFFENSE                                         
ARSON                       38.962474 -76.921305
ASSAULT W/DANGEROUS WEAPON  38.988508 -76.910146
BURGLARY                    38.992170 -76.912290
HOMICIDE                    38.973603 -76.912994
MOTOR VEHICLE THEFT         38.989381 -76.910021
ROBBERY                     38.989706 -76.910146
SEX ABUSE                   38.971907 -76.913666
THEFT F/AUTO                38.993560 -76.910931
THEFT/OTHER                 38.988688 -76.910930

In [58]:
df.groupby('OFFENSE')['X','Y','LATITUDE','LONGITUDE','XCOORD','YCOORD'].max()

X          Y   LATITUDE  LONGITUDE  \
OFFENSE                                                                  
ARSON                      -76.921307  38.962482  38.962474 -76.921305   
ASSAULT W/DANGEROUS WEAPON -76.910148  38.988516  38.988508 -76.910146   
BURGLARY                   -76.912292  38.992178  38.992170 -76.912290   
HOMICIDE                   -76.912996  38.973610  38.973603 -76.912994   
MOTOR VEHICLE THEFT        -76.910023  38.989389  38.989381 -76.910021   
ROBBERY                    -76.910148  38.989714  38.989706 -76.910146   
SEX ABUSE                  -76.913668  38.971915  38.971907 -76.913666   
THEFT F/AUTO               -76.910933  38.993568  38.993560 -76.910931   
THEFT/OTHER                -76.910932  38.988695  38.988688 -76.910930   

                               XCOORD     YCOORD  
OFFENSE                                           
ARSON                       406824.00  143808.19  
ASSAULT W/DANGEROUS WEAPON  407828.70  146678.22  
BURGLARY                    407584.27  147140.35  
HOMICIDE                    407528.83  145092.96  
MOTOR VEHICLE THEFT         407810.36  146861.16  
ROBBERY                     407828.70  146838.57  
SEX ABUSE                   407489.85  144905.56  
THEFT F/AUTO                407763.60  147424.26  
THEFT/OTHER                 407757.28  146749.51